### PR TITLE
feat: add backtraces

### PR DESCRIPTION
**Pull request by doinkythederp**
## Describe the changes this PR makes. Why should it be merged?

Adds backtraces which are exposed through a new `Backtrace` struct.

Backtraces are shown alongside panic messages and can be manually captured using `Backtrace::capture` or `Backtrace::try_capture`. `Backtrace::capture` silently fails if the current environment does not support backtraces, such as in WASM environments or if the user manually disables including unwind tables with `RUSTFLAGS="-Cforce-unwind-tables=off"`.

## Additional Context
- I have tested these changes on a VEX V5 brain.
- These changes update the crate's interface (e.g. functions/modules added or changed).

Backtrace capturing is backed by LLVM libunwind and the resulting frames are instruction addresses that can be symbolized with `llvm-symbolizer` or `addr2line`.

Adding the libunwind library adds about 12k to vexide, bringing the basic example up to 62k. The unwind tables themselves are low-impact:

| No libunwind, no unwind tables | No unwind tables | Backtrace-ready |
|--------------------------------|------------------|-----------------|
| 50k | 62k | 63k |

This change is compatible with the previous version of the armv7a-vex-v5 target file, but backtraces will fail to capture frames.

### CODE REVIEW

**Review by Tropix126 — CHANGES_REQUESTED** [review]

**Comment by Tropix126** [comment]
I also think that we should consider putting this into an optional feature.

File: armv7a-vex-v5.json [hunk 1]
```diff
@@ -13,15 +13,11 @@
     "max-atomic-width": 64,
     "panic-strategy": "abort",
     "post-link-args": {
-        "ld.lld": [
```
[the remark this answers is not in the document]
Comment (doinkythederp) [reply armv7a-vex-v5.json:16]: This was actually just a reformatting change and I haven't changed the functionality of the target file's post-link-args.

File: packages/vexide-core/src/backtrace.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        while cursor.step()? {
+            let mut instruction_pointer = cursor.register(registers::UNW_REG_IP)?;
+
+            // Adjust the IP to point within the function symbol. This should
+            // only be done if the frame is not a signal frame.
+
+            // TODO before merging: does this really affect anything?
+            if !cursor.is_signal_frame()? {
+                instruction_pointer -= 1;
+            }
```
Comment (doinkythederp) [inline packages/vexide-core/src/backtrace.rs:84]: This is mostly ready but I need to do some hardware testing to see if this code affects backtrace quality.

**Comment by doinkythederp** [comment]
`unw_resume` in vex-libunwind should probably have the C-unwind abi

File: armv7a-vex-v5.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (Gavin-Niederman) [reply armv7a-vex-v5.json:16]: > any changes to this file should be done upstream to cargo-v5

^^ this

[the remark this answers is not in the document]
Comment (doinkythederp) [reply armv7a-vex-v5.json:16]: > > any changes to this file should be done upstream to cargo-v5
> 
> ^^ this

That’s out of scope in this PR - if it’s merged, that would be the next step. But I can’t change cargo-v5 files from this PR and there’s no reason to update the cargo-v5 target file unless this PR is merged.

File: packages/vexide-core/Cargo.toml [hunk 3]
```diff
@@ -44,4 +47,6 @@ libm = "0.2.8"
 workspace = true
 
 [features]
+default = ["unwind"]
```
Comment (Tropix126) [inline packages/vexide-core/Cargo.toml:50]: I think that this should be called `backtrace` or similar until (or if) we get proper unwinding figured out.

**Review by Tropix126 — APPROVED** [review]
LGTM otherwise

**Review by Gavin-Niederman — APPROVED** [review]
LGTM

File: packages/vexide-core/Cargo.toml (hunk 3 above)
Comment (doinkythederp) [reply packages/vexide-core/Cargo.toml:50]: That makes sense - I named this after the dependency (libunwind), but it's better for users for it to be named it after what it does (backtraces)

**Comment by doinkythederp** [comment]
BTW I ran a formatter on the Cargo.toml files

**Pull request merged**